src/Provider/AbstractProvider.php

```diff
@@ -16,64 +16,44 @@
 abstract class AbstractProvider implements ProviderInterface
 {
     /**
-     * @var string
+     * @var string HTTP method used to fetch access tokens.
      */
-    public $clientId = '';
+    const ACCESS_TOKEN_METHOD = 'post';
 
     /**
-     * @var string
+     * @var string Key used in the access token response to identify the user.
      */
-    public $clientSecret = '';
+    const ACCESS_TOKEN_UID = 'uid';
 
     /**
-     * @var string
+     * @var string Type of response expected from the provider.
      */
-    public $redirectUri = '';
+    const RESPONSE_TYPE = 'json';
```

Comment:
Having the response type as a constant is problematic for the Facebook provider. In Graph v2.2 and below the response type was "string" and starting in v2.3 the response type was "json". See how the [Facebook provider was dealing with this issue](https://github.com/thephpleague/oauth2-facebook/blob/master/src/Provider/Facebook.php#L130-L144). Any way we can revert this change for the response type?

Also - although I like constants, there's no guarantee that any of these constants will be consistent across all versions of a provider's API (as in the response type example). Perhaps we should revert all of them to be on the safe side? :)